[Errno 13] Permission denied error on Fedora 36

Fedora 36 with all updates applied. Selecting a document and then pressing Convert shows an error "[Errno 13] Permission denied: '/tmp/input_file'"  I have tried opening the file from a different location and renaming the file but the issue remains.

Running from terminal shows the following log:

`[fhit@fedora /]$ dangerzone
QSocketNotifier: Can only be used with threads started with QThread
QObject::connect: No such signal QPlatformNativeInterface::systemTrayWindowChanged(QScreen*)
[INFO] > /usr/bin/podman run --network none -u dangerzone --security-opt no-new-privileges --userns keep-id --cap-drop all --rm -v /home/fhit/Downloads/VFST.pdf:/tmp/input_file -v /home/fhit/.config/dangerzone/tmp/tmptdfvm10l/pixels:/dangerzone dangerzone.rocks/dangerzone /usr/bin/python3 /usr/local/bin/dangerzone.py document-to-pixels
[ERROR] Invalid JSON returned from container: Traceback (most recent call last):

[ERROR] Invalid JSON returned from container:   File "/usr/local/bin/dangerzone.py", line 542, in <module>

[ERROR] Invalid JSON returned from container:     sys.exit(main())

[ERROR] Invalid JSON returned from container:   File "/usr/local/bin/dangerzone.py", line 533, in main

[ERROR] Invalid JSON returned from container:     return converter.document_to_pixels()

[ERROR] Invalid JSON returned from container:   File "/usr/local/bin/dangerzone.py", line 105, in document_to_pixels

[ERROR] Invalid JSON returned from container:     mime_type = mime.from_file("/tmp/input_file")

[ERROR] Invalid JSON returned from container:   File "/usr/lib/python3.10/site-packages/magic/__init__.py", line 111, in from_file

[ERROR] Invalid JSON returned from container:     with _real_open(filename):

[ERROR] Invalid JSON returned from container: PermissionError: [Errno 13] Permission denied: '/tmp/input_file'

[ERROR] documents-to-pixels failed`

--- a/dangerzone/isolation_provider/container.py
+++ b/dangerzone/isolation_provider/container.py
@@ -220,10 +220,13 @@ class Container(IsolationProvider):
         # Create a temporary directory inside the cache directory for this run. Then,
         # create some subdirectories for the various stages of the file conversion:
         #
+        # * unsafe: Where the input file will be copied
         # * pixel: Where the RGB data will be stored
         # * safe: Where the final PDF file will be stored
         with tempfile.TemporaryDirectory(dir=get_tmp_dir()) as t:
             tmp_dir = pathlib.Path(t)
+            unsafe_dir = tmp_dir / "unsafe"
+            unsafe_dir.mkdir()
             pixel_dir = tmp_dir / "pixels"
             pixel_dir.mkdir()
             safe_dir = tmp_dir / "safe"
@@ -231,6 +234,7 @@ class Container(IsolationProvider):
 
             return self._convert_with_tmpdirs(
                 document=document,
+                unsafe_dir=unsafe_dir,
                 pixel_dir=pixel_dir,
                 safe_dir=safe_dir,
                 ocr_lang=ocr_lang,
@@ -240,6 +244,7 @@ class Container(IsolationProvider):
     def _convert_with_tmpdirs(
         self,
         document: Document,
+        unsafe_dir: pathlib.Path,
         pixel_dir: pathlib.Path,
         safe_dir: pathlib.Path,
         ocr_lang: Optional[str],
@@ -252,6 +257,9 @@ class Container(IsolationProvider):
         else:
             ocr = "0"
 
+        copied_file = unsafe_dir / "input_file"
+        shutil.copyfile(f"{document.input_filename}", copied_file)
+
         # Convert document to pixels
         command = [
             "/usr/bin/python3",
@@ -260,7 +268,7 @@ class Container(IsolationProvider):
         ]
         extra_args = [
             "-v",
-            f"{document.input_filename}:/tmp/input_file:Z",
+            f"{copied_file}:/tmp/input_file:Z",
             "-v",
             f"{pixel_dir}:/dangerzone:Z",
             "-e",

--- a/dangerzone/isolation_provider/container.py
+++ b/dangerzone/isolation_provider/container.py
@@ -9,10 +9,8 @@ import subprocess
 import tempfile
 from typing import Callable, List, Optional, Tuple
 
-import appdirs
-
 from ..document import Document
-from ..util import get_resource_path, get_subprocess_startupinfo
+from ..util import get_resource_path, get_subprocess_startupinfo, get_tmp_dir
 from .base import IsolationProvider
 
 # Define startupinfo for subprocesses
@@ -225,9 +223,7 @@ class Container(IsolationProvider):
         else:
             ocr = "0"
 
-        dz_tmp = os.path.join(appdirs.user_config_dir("dangerzone"), "tmp")
-        os.makedirs(dz_tmp, exist_ok=True)
-
+        dz_tmp = get_tmp_dir()
         tmpdir = tempfile.TemporaryDirectory(dir=dz_tmp)
         pixel_dir = os.path.join(tmpdir.name, "pixels")
         safe_dir = os.path.join(tmpdir.name, "safe")

--- a/dangerzone/isolation_provider/container.py
+++ b/dangerzone/isolation_provider/container.py
@@ -260,9 +260,9 @@ class Container(IsolationProvider):
         ]
         extra_args = [
             "-v",
-            f"{document.input_filename}:/tmp/input_file",
+            f"{document.input_filename}:/tmp/input_file:Z",
             "-v",
-            f"{pixel_dir}:/dangerzone",
+            f"{pixel_dir}:/dangerzone:Z",
             "-e",
             f"ENABLE_TIMEOUTS={self.enable_timeouts}",
         ]
@@ -280,9 +280,9 @@ class Container(IsolationProvider):
             ]
             extra_args = [
                 "-v",
-                f"{pixel_dir}:/dangerzone",
+                f"{pixel_dir}:/dangerzone:Z",
                 "-v",
-                f"{safe_dir}:/safezone",
+                f"{safe_dir}:/safezone:Z",
                 "-e",
                 f"OCR={ocr}",
                 "-e",

--- a/dangerzone/isolation_provider/container.py
+++ b/dangerzone/isolation_provider/container.py
@@ -2,6 +2,7 @@ import gzip
 import json
 import logging
 import os
+import pathlib
 import pipes
 import platform
 import shutil
@@ -216,19 +217,40 @@ class Container(IsolationProvider):
         ocr_lang: Optional[str],
         stdout_callback: Optional[Callable] = None,
     ) -> bool:
+        # Create a temporary directory inside the cache directory for this run. Then,
+        # create some subdirectories for the various stages of the file conversion:
+        #
+        # * pixel: Where the RGB data will be stored
+        # * safe: Where the final PDF file will be stored
+        with tempfile.TemporaryDirectory(dir=get_tmp_dir()) as t:
+            tmp_dir = pathlib.Path(t)
+            pixel_dir = tmp_dir / "pixels"
+            pixel_dir.mkdir()
+            safe_dir = tmp_dir / "safe"
+            safe_dir.mkdir()
+
+            return self._convert_with_tmpdirs(
+                document=document,
+                pixel_dir=pixel_dir,
+                safe_dir=safe_dir,
+                ocr_lang=ocr_lang,
+                stdout_callback=stdout_callback,
+            )
+
+    def _convert_with_tmpdirs(
+        self,
+        document: Document,
+        pixel_dir: pathlib.Path,
+        safe_dir: pathlib.Path,
+        ocr_lang: Optional[str],
+        stdout_callback: Optional[Callable] = None,
+    ) -> bool:
         success = False
 
         if ocr_lang:
             ocr = "1"
         else:
             ocr = "0"
-
-        dz_tmp = get_tmp_dir()
-        tmpdir = tempfile.TemporaryDirectory(dir=dz_tmp)
-        pixel_dir = os.path.join(tmpdir.name, "pixels")
-        safe_dir = os.path.join(tmpdir.name, "safe")
-        os.makedirs(pixel_dir, exist_ok=True)
-        os.makedirs(safe_dir, exist_ok=True)
 
         # Convert document to pixels
         command = [
@@ -283,9 +305,6 @@ class Container(IsolationProvider):
 
                 # We did it
                 success = True
-
-        # Clean up
-        tmpdir.cleanup()
 
         return success
 

--- a/dangerzone/logic.py
+++ b/dangerzone/logic.py
@@ -9,10 +9,9 @@ import subprocess
 import sys
 from typing import Callable, List, Optional
 
-import appdirs
 import colorama
 
-from . import errors
+from . import errors, util
 from .document import Document
 from .isolation_provider.base import IsolationProvider
 from .settings import Settings
@@ -31,7 +30,7 @@ class DangerzoneCore(object):
         colorama.init(autoreset=True)
 
         # App data folder
-        self.appdata_path = appdirs.user_config_dir("dangerzone")
+        self.appdata_path = util.get_config_dir()
 
         # Languages supported by tesseract
         with open(get_resource_path("ocr-languages.json"), "r") as f:

--- a/dangerzone/util.py
+++ b/dangerzone/util.py
@@ -2,6 +2,24 @@ import pathlib
 import platform
 import subprocess
 import sys
+from typing import Optional
+
+import appdirs
+
+
+def get_config_dir() -> str:
+    return appdirs.user_config_dir("dangerzone")
+
+
+def get_tmp_dir() -> Optional[str]:
+    """Get the parent dir for the Dangerzone temporary dirs.
+
+    This function returns the parent directory where Dangerzone will store its temporary
+    directories. The default behavior is to let Python choose for us (e.g., in `/tmp`
+    for Linux), which is why we return None. However, we still need to define this
+    function in order to be able to set this dir via mocking in our tests.
+    """
+    return None
 
 
 def get_resource_path(filename: str) -> str:

--- a/tests/test_cli.py
+++ b/tests/test_cli.py
@@ -10,6 +10,7 @@ import tempfile
 import traceback
 from pathlib import Path
 from typing import Sequence
+from unittest import mock
 
 import pytest
 from click.testing import CliRunner, Result
@@ -129,18 +130,25 @@ class TestCli(TestBase):
         if os.environ.get("DUMMY_CONVERSION", False):
             args = ("--unsafe-dummy-conversion", *args)
 
-        # TODO: Replace this with `contextlib.chdir()` [1], which was added in
-        # Python 3.11.
-        #
-        # [1]: # https://docs.python.org/3/library/contextlib.html#contextlib.chdir
-        try:
-            if tmp_path is not None:
-                cwd = os.getcwd()
-                os.chdir(tmp_path)
-            result = CliRunner().invoke(cli_main, args)
-        finally:
-            if tmp_path is not None:
-                os.chdir(cwd)
+        with tempfile.TemporaryDirectory() as t:
+            tmp_dir = Path(t)
+            # TODO: Replace this with `contextlib.chdir()` [1], which was added in
+            # Python 3.11.
+            #
+            # [1]: # https://docs.python.org/3/library/contextlib.html#contextlib.chdir
+            try:
+                if tmp_path is not None:
+                    cwd = os.getcwd()
+                    os.chdir(tmp_path)
+
+                with mock.patch(
+                    "dangerzone.isolation_provider.container.get_tmp_dir",
+                    return_value=t,
+                ):
+                    result = CliRunner().invoke(cli_main, args)
+            finally:
+                if tmp_path is not None:
+                    os.chdir(cwd)
 
         return CLIResult.reclass_click_result(result, args)
 

--- a/tests/test_cli.py
+++ b/tests/test_cli.py
@@ -120,7 +120,7 @@ class TestCli(TestBase):
         (i.e., perform `shlex.split()`), as this is prone to errors in Windows
         environments [1]. The user must perform the tokenizaton themselves.
 
-        [1]: https://stackoverflow.com/a/35900070c
+        [1]: https://stackoverflow.com/a/35900070
         """
         if isinstance(args, str):
             # Convert the single argument to a tuple, else Click will attempt
@@ -135,7 +135,7 @@ class TestCli(TestBase):
             # TODO: Replace this with `contextlib.chdir()` [1], which was added in
             # Python 3.11.
             #
-            # [1]: # https://docs.python.org/3/library/contextlib.html#contextlib.chdir
+            # [1]: https://docs.python.org/3/library/contextlib.html#contextlib.chdir
             try:
                 if tmp_path is not None:
                     cwd = os.getcwd()

--- a/tests/test_cli.py
+++ b/tests/test_cli.py
@@ -150,6 +150,10 @@ class TestCli(TestBase):
                 if tmp_path is not None:
                     os.chdir(cwd)
 
+                if tmp_dir.exists():
+                    stale_files = list(tmp_dir.iterdir())
+                    assert not stale_files
+
         return CLIResult.reclass_click_result(result, args)
 
 

